util/workers/worker.go

```diff
@@ -0,0 +1,200 @@
+// Copyright 2023 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package workers
+
+import (
+	"context"
+	"math"
+	"sync"
+	"sync/atomic"
+)
+
+const (
+	// MaxWorker is the max number of worker.
+	MaxWorker = math.MaxUint16
+)
+
+// Worker interface.
+type Worker[task any] interface {
+	HandleTask(context.Context, task) error
+	Close()
+}
+
+// NewWorker creates a Worker.
+type NewWorker[task any, w Worker[task]] func() w
+
+// WorkerManager is the pool.
+type WorkerManager[task any, w Worker[task]] struct {
+	max    int32
+	used   atomic.Int32
+	count  atomic.Int32
+	closed atomic.Bool
+	flying struct {
+		*sync.Cond
+		atomic.Int64
+	}
+	ctx       context.Context
+	cancel    context.CancelFunc
+	newWorker NewWorker[task, w]
+	workers   []w
+	ch        chan task
+	closeCh   chan struct{}
+	wg        sync.WaitGroup
+}
+
+// New creates WorkerManager.
+// The last input parameter _ *task is used for type inference, you need to use (**TypeOfYourTask)(nil).
+func New[task any, w Worker[task]](ctx context.Context, least, max int, newWorker NewWorker[task, w], _ *task) *WorkerManager[task, w] {
+	if max == 0 {
+		return nil
+	}
+	if max > MaxWorker {
+		max = MaxWorker
+	}
+	if least > max {
+		least = max
+	}
+	ctx1, cancel := context.WithCancel(ctx)
+	worker := &WorkerManager[task, w]{
+		max: int32(max),
+		flying: struct {
+			*sync.Cond
+			atomic.Int64
+		}{
+			Cond: sync.NewCond(&sync.Mutex{}),
+		},
+		ctx:       ctx1,
+		cancel:    cancel,
+		newWorker: newWorker,
+		workers:   make([]w, max),
+		ch:        make(chan task, max), // TODO: unbounded channel.
```

Comment:
No need to be unbounded, a buffered one is enough.
And even the buffer size doesn't need to be big, because as long as the worker consumes fast enough, the task shouldn't piled up.

If the task really piled up here, it's more likely the implementation is buggy.

Reply:
Removed this TODO, and set the buffer size to max concurrency to avoid worker waiting for tasks.